File: 46-permutations/leochang/46.cpp
class Solution {
public:
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> ans;
        recursive(nums, ans, 0);
        return ans;
    }
    void recursive(vector<int> &nums, vector<vector<int>> &ans, int idx) {
        if (idx == nums.size()) {
            ans.push_back(nums);
        }
        for (int i = idx; i < nums.size(); i++) {
            swap(nums[idx], nums[i]);
            recursive(nums, ans, idx+1);
            swap(nums[idx], nums[i]);
        }
    }
};

File: 46-permutations/ken/46.py
class Solution(object):
    def permute(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """

        if len(nums) == 1:
            return [list(nums)]

        ans = []

        for i in xrange(len(nums)):
            n = nums[i]
            del nums[i]

            tmp = self.permute(nums)

            for t in tmp:
                t.insert(0, n)
                ans.append(t)

            nums.insert(i, n)

        return ans
